[FLINK-13360][documentation] Add documentation for HBase connector for Table API & SQL (#9799)
apache/flink: docs/dev/table/connect.md

| Apache Kafka      | 0.10                | `flink-connector-kafka-0.10` | [Download](http://central.maven.org/maven2/org/apache/flink/flink-sql-connector-kafka-0.10{{site.scala_version_suffix}}/{{site.version}}/flink-sql-connector-kafka-0.10{{site.scala_version_suffix}}-{{site.version}}.jar) |
| Apache Kafka      | 0.11                | `flink-connector-kafka-0.11` | [Download](http://central.maven.org/maven2/org/apache/flink/flink-sql-connector-kafka-0.11{{site.scala_version_suffix}}/{{site.version}}/flink-sql-connector-kafka-0.11{{site.scala_version_suffix}}-{{site.version}}.jar) |
| Apache Kafka      | 0.11+ (`universal`) | `flink-connector-kafka`      | [Download](http://central.maven.org/maven2/org/apache/flink/flink-sql-connector-kafka{{site.scala_version_suffix}}/{{site.version}}/flink-sql-connector-kafka{{site.scala_version_suffix}}-{{site.version}}.jar) |

> | HBase             | 1.4.3               | `flink-hbase`                | [Download](http://central.maven.org/maven2/org/apache/flink/flink-hbase{{site.scala_version_suffix}}/{{site.version}}/flink-hbase{{site.scala_version_suffix}}-{{site.version}}.jar) |

Reviewer: I'm not sure about this. From my understanding, the downloaded jar should contain **shaded** HBase dependencies. So, we may need a module called `flink-sql-connector-hbase` to shade HBase.

Author: Yeah, you are right, More appropriate way is use shade HBase. I created a JIRA to track this: https://issues.apache.org/jira/browse/FLINK-14359

{% top %}

### HBase Connector

<span class="label label-primary">Source: Batch</span>
<span class="label label-primary">Sink: Batch</span>
<span class="label label-primary">Sink: Streaming Append Mode</span>
<span class="label label-primary">Sink: Streaming Upsert Mode</span>
<span class="label label-primary">Temporal Join: Sync Mode</span>

The HBase connector allows for reading from and writing to an HBase cluster.

The connector can operate in [upsert mode](#update-modes) for exchanging UPSERT/DELETE messages with the external system using a [key defined by the query](./streaming/dynamic_tables.html#table-to-stream-conversion).

For append-only queries, the connector can also operate in [append mode](#update-modes) for exchanging only INSERT messages with the external system.

To use this connector, add the following dependency to your project:

{% highlight xml %}
<dependency>
  <groupId>org.apache.flink</groupId>
  <artifactId>flink-connector-hbase{{ site.scala_version_suffix }}</artifactId>
  <version>{{ site.version }}</version>
</dependency>
{% endhighlight %}

The connector can be defined as follows:

> <div data-lang="DDL" markdown="1">

Reviewer: missing `<div class="codetabs" markdown="1">` above this line.

{% highlight sql %}
CREATE TABLE MyUserTable (
  hbase_rowkey_name rowkey_type,
  hbase_column_family_name1 ROW<...>,
  hbase_column_family_name2 ROW<...>
) WITH (
  'connector.type' = 'hbase', -- required: specify this table type is hbase
  
  'connector.version' = '1.4.3',          -- required: valid connector versions are "1.4.3"
  
  'connector.table-name' = 'hbase_table_name',  -- required: hbase table name
  

>   'connector.zookeeper.quorum' = 'quorum_url', -- required: hbase zookeeper config

Reviewer: ```suggestion
  'connector.zookeeper.quorum' = 'localhost:2181', -- required: HBase Zookeeper quorum configuration
```
I would suggest to use a meaningful example value.

>   'connector.zookeeper.znode.parent' = 'znode',

Reviewer: ```suggestion
  'connector.zookeeper.znode.parent' = '/test',  -- rquired: the root dir in Zookeeper for HBase cluster
```

The same above.

  'connector.write.buffer-flush.max-size' = '1048576', -- optional: Write option, sets when to flush a buffered request
                                                       -- based on the memory size of rows currently added.

  'connector.write.buffer-flush.max-rows' = '1', -- optional: Write option, sets when to flush buffered 
                                                    -- request based on the number of rows currently added.

  'connector.write.buffer-flush.interval' = '1', -- optional: Write option, sets a flush interval flushing buffered 
                                                 -- requesting if the interval passes, in milliseconds.
)
{% endhighlight %}
</div>
</div>

> **Column family:** Values other than `rowKey` must be declared as column families, and all column family values must be wrapped with the SQL ROW function before being inserted into HBase table.

Reviewer: ```suggestion
**Columns:** All the column families in HBase table must be declared as `ROW` type, the field name maps to the column family name, and the nested field names map to the column qualifier names. There is no need to declare all the families and qualifiers in the schema, users can declare what's necessary. Except the `ROW` type fields, the only one field of atomic type (e.g. `STRING`, `BIGINT`) will be recognized as row key of the table. There's no constraints on the name of row key field. 
```

I rephrased this sentence. What do you think about it?

Author: Good~

**HBase config:** If need to configure Config for HBase, create default configuration from current runtime env (`hbase-site.xml` in classpath) first, and overwrite configuration using serialized configuration from client-side env (`hbase-site.xml` in classpath).

**Temporary join:** Lookup join against HBase do not use any caching; data is always queired directly through the HBase client.

**Rowkey:** Empty string rowkey values are currently unsupported.

> 

Reviewer: **Java/Scala/Python API:** Java/Scala/Python APIs are not supported yet.

{% top %}

### HBase Connector

<span class="label label-primary">Source: Batch</span>
<span class="label label-primary">Sink: Batch</span>
<span class="label label-primary">Sink: Streaming Append Mode</span>
<span class="label label-primary">Sink: Streaming Upsert Mode</span>
<span class="label label-primary">Temporal Join: Sync Mode</span>

The HBase connector allows for reading from an HBase cluster.

> The HBase connector allows for writing into an HBase cluster.

Reviewer 2: This can be one sentence; 

The HBase connector allows for reading from and writing to an HBase cluster.

> The connector can operate in [upsert mode](#update-modes) for exchanging UPSERT/DELETE messages with the external system using a [key defined by the query](./streaming/dynamic_tables.html#table-to-stream-conversion).

Reviewer 2: ```suggestion
The connector can operate in [upsert mode](#update-modes) by executing UPSERT and DELETE statements on the cluster using a [key defined by the query](./streaming/dynamic_tables.html#table-to-stream-conversion).
```

Author: This is same to Elasticsearch, it is messages instead of statements.

> For append-only queries, the connector can also operate in [append mode](#update-modes) for exchanging only INSERT messages with the external system.

Reviewer 2: ```suggestion
For append-only queries, the connector can also operate in [append mode](#update-modes) by exeucting INSERT statements.
```

Author: Ditto

To use this connector, add the following dependency to your project:

{% highlight xml %}
<dependency>
  <groupId>org.apache.flink</groupId>
  <artifactId>flink-connector-hbase{{ site.scala_version_suffix }}</artifactId>

>   <version>{{site.version }}</version>

Reviewer 2: nit: 
```suggestion
  <version>{{ site.version }}</version>
```

</dependency>
{% endhighlight %}

The connector can be defined as follows:

<div data-lang="DDL" markdown="1">
{% highlight sql %}
CREATE TABLE MyUserTable (
  hbase_rowkey_name rowkey_type,
  hbase_column_family_name1 ROW<...>,
  hbase_column_family_name2 ROW<...>
) WITH (

>   'connector.type' = 'hbase', -- required: specify this table type is hbase

Reviewer 2: I think the DDL properties would be much clearer as a table and only show create table statements in examples. @twalthr and @dawidwys what do you think?

Reviewer 2: If so this is a change we would want to make everywhere.

Author: not support, now, we don't have connector api for HBase and JDBC.

Reviewer 3: I've filed the issue: https://issues.apache.org/jira/browse/FLINK-14349  to address the HBase Java/Scala API, Will create a PR for it. Thanks.

Author: Thanks @openinx~

  
  'connector.version' = '1.4.3',          -- required: valid connector versions are "1.4.3"
  
  'connector.table-name' = 'hbase_table_name',  -- required: hbase table name
  
  'connector.zookeeper.quorum' = 'quorum_url', -- required: hbase zookeeper config
  'connector.zookeeper.znode.parent' = 'znode',

  'connector.write.buffer-flush.max-size' = '1048576', -- optional: Write option, sets when to flush a buffered request
                                                       -- based on the memory size of rows currently added.

  'connector.write.buffer-flush.max-rows' = '1', -- optional: Write option, sets when to flush buffered 
                                                    -- request based on the number of rows currently added.

  'connector.write.buffer-flush.interval' = '1', -- optional: Write option, sets a flush interval flushing buffered 
                                                 -- requesting if the interval passes, in milliseconds.
)
{% endhighlight %}
</div>
</div>

> **Column family:** Values other than rowKey must be declared by column families. So need to wrap values with the SQL ROW function before inserting into hbase table.

Reviewer 2: ```suggestion
**Column family:** Values other than `rowKey` must be declared as column families, and all column family values must be wrapped with the SQL ROW function before being inserted into HBase table.
```

> **HBase config:** If need to configure Config for HBase, create default configuration from current runtime env (`hbase-site.xml` in classpath) first, and overwrite configuration using serialized configuration from client-side env (`hbase-site.xml` in classpath).

Reviewer 2: I found this sentence very confusing and I'm not entirely sure what you are trying to communicate.

Author: This is cluster config to let user configure HBase client more.

Author: @wuchong Any thought about this?

Reviewer: I would like to remove this sentence. All the client behavior should be configured via WITH properties. Passing `hbase-site.xml` is really hard to use.

Author: Let's remove it.

> **Temporary join:** The Lookup Join of HBase does not use any caching, and every time the data is accessed directly to the client Api of HBase.

Reviewer 2: ```suggestion
**Temporary join:** Lookup join against HBase do not use any caching; data is always queired directly through the HBase client.
```

> **Rowkey:** User should confirm rowkey should not be empty string. (waiting for support)

Reviewer 2: ```suggestion
**Rowkey:** Empty string rowkey values are currently unsupported.
```

                                 # based on the memory size of rows currently added.
    max-rows: 1                  # optional: Write option, sets when to flush buffered 
                                 # request based on the number of rows currently added.
    interval: 1                  # optional: Write option, sets a flush interval flushing buffered 
                                 # requesting if the interval passes, in milliseconds.
{% endhighlight %}
</div>

<div data-lang="DDL" markdown="1">
{% highlight sql %}
CREATE TABLE MyUserTable (
  hbase_rowkey_name rowkey_type,
  hbase_column_family_name1 ROW<...>,
  hbase_column_family_name2 ROW<...>
) WITH (
  'connector.type' = 'hbase', -- required: specify this table type is hbase
  
  'connector.version' = '1.4.3',          -- required: valid connector versions are "1.4.3"
  
  'connector.table-name' = 'hbase_table_name',  -- required: hbase table name
  
  'connector.zookeeper.quorum' = 'localhost:2181', -- required: HBase Zookeeper quorum configuration
  'connector.zookeeper.znode.parent' = '/test',  -- required: the root dir in Zookeeper for HBase cluster

>   'connector.write.buffer-flush.max-size' = '1048576', -- optional: Write option, sets when to flush a buffered request

Reviewer: ```suggestion
  'connector.write.buffer-flush.max-size' = '10mb', -- optional: writing option, determines how many size in memory of buffered rows to insert per round trip. This can help performance on writing to JDBC database. The default value is "2mb".
```

                                                       -- based on the memory size of rows currently added.

>   'connector.write.buffer-flush.max-rows' = '1', -- optional: Write option, sets when to flush buffered 

Reviewer: ```suggestion
  'connector.write.buffer-flush.max-rows' = '1000', -- optional: writing option, determines how many rows to insert per round trip. This can help performance on writing to JDBC database. No default value, i.e. the default flushing is not depends on the number of buffered rows.
```

                                                    -- request based on the number of rows currently added.

>   'connector.write.buffer-flush.interval' = '1', -- optional: Write option, sets a flush interval flushing buffered 

Reviewer: 'connector.write.buffer-flush.interval' = '2s', -- optional: writing option, ............. Default value is "0s", which means no asynchronous flush thread will be scheduled.

    znode.parent: "/test"        # required: the root dir in Zookeeper for HBase cluster
    
  write.buffer-flush:
    max-size: 1048576            # optional: writing option, determines how many size in memory of buffered
                                 # rows to insert per round trip. This can help performance on writing to JDBC
                                 # database. The default value is "2mb".
    max-rows: 1                  # optional: writing option, determines how many rows to insert per round trip.
                                 #This can help performance on writing to JDBC database. No default value,
                                 # i.e. the default flushing is not depends on the number of buffered rows.

>     interval: 1                  # optional: writing option, sets a flush interval flushing buffered requesting

Reviewer: Please also update the values in YAML.

Author: Sorry, forgot this, Updated.
